### PR TITLE
Resolve CodeQL alerts: two polynomial-ReDoS regexes + workflow permissions

### PR DESCRIPTION
**Pull request by akbisw**
## Summary
Addresses the open GitHub **code-scanning (CodeQL)** alerts found while auditing the repo's security tab.

| # | Severity | Location | Rule |
|---|----------|----------|------|
| 4 | HIGH | `client/config.ts` (trailing-slash trim) | `js/polynomial-redos` |
| 3 | HIGH | `client/util.ts` (Lambda error message) | `js/polynomial-redos` |
| 1 | MEDIUM | `.github/workflows/main.yml` | `actions/missing-workflow-permissions` |

## Fixes
- **`config.ts`** — `/\/+$/` backtracks polynomially on a long run of `/`. Replaced with a linear backward scan.
- **`util.ts`** — `/^.+failed with error (.+)$/` (run on a Cognito `UserLambdaValidationException` message) backtracks polynomially on a crafted near-miss. Replaced with `lastIndexOf`/`slice`, mirroring the regex's semantics (non-empty prefix and suffix) with no backtracking.
- **`main.yml`** — added an explicit minimal `permissions: contents: read`. The test job only reads the repo; coverage upload uses a dedicated `CODECOV_TOKEN` secret, not `GITHUB_TOKEN`.

## Test plan
- New `test/redos-fixes.test.ts`: message extraction unchanged (incl. no-marker / no-prefix / no-suffix cases); trailing-slash trim unchanged; both paths stay fast on pathological inputs.
- Full suite: 449 passed / 0 failures, twice; tsc and eslint clean.
- CodeQL re-scan on this PR should clear alerts #3 and #4 (the backtracking regexes are gone) and #1 (permissions block added).

## Note on the Dependabot alert
The one open Dependabot alert — esbuild `< 0.28.1` RCE — is already addressed by the in-flight bump [#90](https://github.com/joinmeow/amazon-cognito-passwordless-auth/pull/90). esbuild is a build-time tool in this repo (the `dist` script), and the advisory's Deno-module vector does not apply to this npm usage; the version bump is still the right remediation.

---

> [!NOTE]
> **Low Risk**
> Behavior-preserving security hardening with targeted tests; CI only narrows default token scope to read-only repo access.
> 
> **Overview**
> Clears CodeQL **js/polynomial-redos** findings by replacing two backtracking regexes with linear string handling, and adds explicit minimal **`permissions: contents: read`** on the test workflow.
> 
> In **`client/config.ts`**, trailing-slash trimming on OAuth/Hosted UI base URLs no longer uses `/\/+$/`; it walks backward from the end of the string so long runs of `/` cannot trigger polynomial backtracking.
> 
> In **`client/util.ts`**, **`UserLambdaValidationException`** message parsing no longer uses `/^.+failed with error (.+)$/`. It uses **`lastIndexOf`** / **`slice`** to pick the rightmost `"failed with error "` with a non-empty prefix and suffix, matching the old greedy-regex behavior without ReDoS risk on crafted Cognito error text.
> 
> **`test/redos-fixes.test.ts`** locks in extraction edge cases, slash normalization, and timing checks on pathological inputs. The CI workflow comment documents that coverage upload uses **`CODECOV_TOKEN`**, not **`GITHUB_TOKEN`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 49c9ab359eecae90cdb42a832b1141c5bb814560. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: client/util.ts [hunk 1]
```diff
@@ -26,9 +26,14 @@ export async function throwIfNot2xx(res: MinimalResponse) {
...
-    if (match) {
-      message = match[1];
+    // Extract the text after the last "failed with error " without the
+    // backtracking regex /^.+failed with error (.+)$/, which is polynomial
+    // on crafted messages (CodeQL js/polynomial-redos). Mirrors the regex:
+    // a non-empty prefix (idx > 0) and a non-empty captured suffix.
+    const marker = "failed with error ";
+    const idx = detail.message.lastIndexOf(marker);
+    if (idx > 0 && idx + marker.length < detail.message.length) {
+      message = detail.message.slice(idx + marker.length);
```
[the remark this answers is not in the document]
Comment (akbisw) [reply client/util.ts]: Fixed in 49c9ab3: the replacement now walks markers right-to-left and uses the rightmost 'failed with error ' that has a non-empty suffix (and non-empty prefix) — exactly what the greedy ^.+ … (.+)$ resolves to (it backtracks past a final suffix-less marker). Verified byte-for-byte equivalent to the original regex across single- and multi-marker cases (incl. 'X failed with error Y failed with error ' → 'Y failed with error '), still O(n) with no backtracking. Regression test added for the multi-marker case.
